File: 576-out-of-boundary-paths/576-out-of-boundary-paths.js
/**
 * @param {number} m
 * @param {number} n
 * @param {number} maxMove
 * @param {number} startRow
 * @param {number} startColumn
 * @return {number}
 */
var findPaths = function(m, n, maxMove, startRow, startColumn) {
    const cache = {};
    const modulus = (10 ** 9) + 7;
    
    const paths = (function dfs(movesLeft, row, col) {
        if (movesLeft < 0) {
            return 0;
        }
        if (row < 0 || col < 0 || row >= m || col >= n) {
            return 1;
        }
        
        const key = JSON.stringify([movesLeft, row, col]);
        
        if (!cache.hasOwnProperty(key)) {
            const val = (
                dfs(movesLeft - 1, row - 1, col) + // up
                dfs(movesLeft - 1, row + 1, col) + // down
                dfs(movesLeft - 1, row, col - 1) + // left
                dfs(movesLeft - 1, row, col + 1)   // right
            );
            cache[key] = val % modulus;
        }
        
        return cache[key];
    })(maxMove, startRow, startColumn);
    
    return paths % modulus;
};

File: 576-out-of-boundary-paths/576-out-of-boundary-paths.py
class Solution:
    def findPaths(self, m: int, n: int, maxMove: int, startRow: int, startColumn: int) -> int:    
        corners = ((0, 0), (0, n - 1), (m - 1, 0), (m - 1, n - 1))
        
        @cache
        def dfs(moves_left: int, row: int, col: int) -> int:
            if moves_left < 0:
                return 0
            if not 0 <= row < m or not 0 <= col < n:
                return 1
            
            return (
                dfs(moves_left - 1, row - 1, col) +
                dfs(moves_left - 1, row, col - 1) +
                dfs(moves_left - 1, row + 1, col) +
                dfs(moves_left - 1, row, col + 1)
            )
        
        return dfs(maxMove, startRow, startColumn) % ((10 ** 9) + 7)